WIP Refresh network policy concepts and application (#33350)
openshift/openshift-docs: networking/network_policy/about-network-policy.adoc

Your cluster must use a network provider that supports the use of network policy, such as OVN-Kubernetes or OpenShift SDN with network policy isolation mode configured.

[NOTE]
====
When using the OpenShift SDN cluster network provider, the following limitations apply regarding network policies:

* Egress network policy as specified by the `egress` field is not supported.
* IPBlock is supported by network policy, but without support for `except` clauses. If you create a policy with an IPBlock section that includes an `except` clause, the SDN pods log warnings and the entire IPBlock section of that policy is ignored.
====

By default, the following roles can create, edit, and delete a network policy:

* `cluster-admin`
* `admin`

[id="limitations-of-network-policy_{context}"]
== Limitations of network policy

The following limitations apply to the use of network policy:

* You cannot use DNS names with network policy.
* You cannot use network policy with host network traffic. Pods with host networking enabled, such as in the following example, are unaffected by network policy rules.
+
[source,yaml]
----
apiVersion: v1
kind: Pod
metadata:
  name: host-net-pod
spec:

>   hostNetwork: true

Reviewer: I'll omit this.

  - podSelector:
      matchLabels:
        deployment: blue
    namespaceSelector:
      matchLabels:
ifeval::[{product-version} < 4.8]
        networking/namespace: web
endif::[]
ifeval::[{product-version} >= 4.8]
        kubernetes.io/metadata.name: web
endif::[]
  policyTypes:
  - Ingress
----

ifeval::[{product-version} < 4.8]
[NOTE]
====
A namespace does not have a label by default. To add a label to a namespace as a target for a namespace selector, add a label such as with the `oc label namespace <name> networking/namespace=<name>` command. Replace `<name>` with the name of the namespace the label is applied to.
====
endif::[]
ifeval::[{product-version} >= 4.8]
[NOTE]
====
Beginning with {product-title} 4.8, which is based on Kubernetes 1.21, the following label is added automatically to each namespace:

[source,yaml]
----
metadata:
  name: <name>
  labels:
    kubernetes.io/metadata.name: <name>
----
====
endif::[]

[id="available-selector-policy-groups_{context}"]
=== Available selector policy groups

Several namespace selectors are available in {product-title}:

`network.openshift.io/policy-group: ingress`::
+
--
Selects all Ingress Controller pods.

[NOTE]
====

> If your `default` Ingress Controller runs on the host network, this namespace selector also allows all traffic for the host network. You can determine if your `default` Ingress Controller uses the `HostNetwork` endpoint publishing strategy with the following command:

Reviewer 2: ++ I said this wrong thanks :)

  --output jsonpath='{.status.endpointPublishingStrategy.type}'
----

If the previous command reports the endpoint publishing strategy as `HostNetwork`, then your `default` Ingress Controller is on the host network.
====
--

`network.openshift.io/policy-group: monitoring`::
Selects all monitoring pods.

`policy-group.network.openshift.io/host-network`: ""::
Selects all host network traffic. You cannot combine this namespace selector with a pod selector.

[id="policy-rules_{context}"]
== Policy rules

A network policy must specify one or more policy rules that identifies what ingress or egress traffic is permitted for pods that the network policy isolates.

[id="policy-rule-logical-ordering_{context}"]
=== Logical ordering

In a single network policy one or more policy rules can be defined. The following logic applies when evaluating an array of one or more ingress or egress policy rules:

* If a policy rule includes both a namespace selector and a pod selector then a logical `and` is used when evaluating both.
* If multiple policy rules are defined then a logical `or` is used when evaluating each policy rule.

[id="policy-rule-precedence_{context}"]
=== Precedence

> Any network policy rule that denies either ingress or egress traffic for a pod is always superseded by a policy rule that allows the denied traffic.

Reviewer 3: Deny everything does not come by default by virtue of applying 'a' policy to control traffic to pod. The traffic needs to blocked explicitly with default-deny policy and then poke holes where ever needed.

  - ports:
    - port: 443
---
kind: NetworkPolicy
apiVersion: networking.k8s.io/v1
metadata:
  name: allow-only-db-https
spec:
  podSelector: {}
  ingress:
    from:
    - namespaceSelector:
        matchLabels:
          app: db
      ports:
      - port: 443
----
////

[id="policy-rule-behavior-with-an-empty-ingress-or-egress-rule-set_{context}"]
=== Behavior with an empty ingress or egress rule set

If you define a network policy rule with an empty list, the network policy does allow any traffic. Any pods selected by the network policy are completely isolated.

For example, the following ingress rule set selects no ingress traffic:

.Example ingress ruleset defined as an empty list
[source,yaml]
----
kind: NetworkPolicy
apiVersion: networking.k8s.io/v1
metadata:
  name: <name>
spec:
  ingress: []
----

////
Name:         name1
Namespace:    default
Created on:   2021-06-28 21:11:24 -0400 EDT
Labels:       <none>
Annotations:  <none>
Spec:
  PodSelector:     <none> (Allowing the specific traffic to all pods in this namespace)
  Allowing ingress traffic:
    <none> (Selected pods are isolated for ingress connectivity)
  Not affecting egress traffic
  Policy Types: Ingress
////

> An empty rule set is the inverse of a rule set with a single empty selector, which selects all traffic. For example, the following ingress rule set selects all ingress traffic:

Reviewer 2: ** which selects and allows all traffic

[source,yaml]
----
kind: NetworkPolicy # Valid
apiVersion: networking.k8s.io/v1
spec:
  ingress:
  - {}
----

////
Spec:
  PodSelector:     <none> (Allowing the specific traffic to all pods in this namespace)
  Allowing ingress traffic:
    To Port: <any> (traffic allowed to all ports)
    From: <any> (traffic not restricted by source)
  Not affecting egress traffic
  Policy Types: Ingress
////

The previous example is identical in behavior to the following example:

.Example ingress ruleset that matches all ingress traffic
[source,yaml]
----
kind: NetworkPolicy # Valid
apiVersion: networking.k8s.io/v1
spec:
  ingress:
  - from:
    - podSelector: {}
      namespaceSelector: {}
----

////
Spec:
  PodSelector:     <none> (Allowing the specific traffic to all pods in this namespace)
  Allowing ingress traffic:
    To Port: <any> (traffic allowed to all ports)
    From:
      NamespaceSelector: <none>
      PodSelector: <none>
  Not affecting egress traffic
  Policy Types: Ingress
////

[id="nework-policy-api_{context}"]
== Network policy API

When you specify network policy, you use the `NetworkPolicy` API in the `networking.k8s.io` API group.

> TODO: Various network policy API tables.

Reviewer 2: network policy api tables?

Author: @astoycos, oh, yeah, so this is probably going to be all the valid fields for the NP API.

toc::[]

> As a project administrator, you can define network policies that restrict traffic to or from pods in a namespace.

Reviewer 4: Not sure the "in a namespace" qualifier is correct here; NetworkPolicy isn't inherently namespace specific.

[NOTE]
====
Any network policies defined by a cluster administrator can be overridden by a project administrator. Do not use network policies to define cluster-wide network traffic policy.
====

What follows is a discussion of how network traffic isolation works with network policy in {product-title}.

[id="network-policy-concepts_{context}"]
== Network policy concepts

You can use a network policy to allow and deny network traffic within a particular namespace. The following concepts apply to network policy:

* Network policy applies to pods in the namespace in which the policy is defined.
* Network policy defines two types of rules:

> ** Ingress rules that allow traffic to pods from other pods or IP address blocks.

Reviewer 2: ```suggestion
** Ingress rules that allow traffic to other pods, Namespaces, or IP address blocks.
```

> ** Egress rules that allow traffic from pods to other pods or IP address blocks.

Reviewer 2: ```suggestion
** Egress rules that allow traffic from other pods, Namespaces, or IP address blocks.
```

* Network policy supports the TCP, UDP, and SCTP protocols.
* Network policy supports specifying ports for supported protocols.

> 

Reviewer 2: I think you need to start with 
1. Who Netpol was designed for (i.e the developer) 
2. Network Policy follows a whitelist model, meaning all traffic to selected pods is denied unless otherwise allowed (Point to Network Policy isolation section) 
3. etc etc.

> A network policy consists of three parts:

Reviewer 2: If anything I would remove the "Network Policy Concepts" header and add one here titled "The Network Policy Object"

Reviewer 2: And I think this section should simply go under the sample Network Policy Object  (similar to what's done in [upstream docs](https://kubernetes.io/docs/concepts/services-networking/network-policies/#networkpolicy-resource))

> * Pod selector: Identifies which pods the network policy isolates. A pod selected by a pod selector is considered to be an _isolated pod_.

Reviewer 2: ```suggestion
* Pod selector: Identifies which pods the network policy applies to. A pod selected by a network policy is considered to be an _isolated pod_.
```

Reviewer 2: This is also proceeding the "Network policy pod isolation behavior" Section which could be a bit confusing

> * Pod selector: Identifies which pods the network policy isolates. A pod selected by a pod selector is considered to be an _isolated pod_.

Reviewer 2: We should also define what "empty" and "nill" means for each of these sections 

i.e "An empty podSelector selects all pods in the namespace."

> * Ingress and egress rules: Identifies _network policy peers_. A peer is a pod selected by an ingress or egress rule selector.

Reviewer 2: ```suggestion
* Ingress and egress rules: Contains an additive list of  _network policy peers_ which define where traffic is allowed from/to.
```
